DietPi-Update | Pre-patch, code and wording

MichaIng opened the pull request:
**Status**: Ready
🈯️ Jessie
🈯️ Stretch

**Commit list/description**:
+ DietPi-Update | Allow RC version updates on all branches
  - For simplification and allows us theoretically to apply those on master branch as well.
  - We have full control via: `server_version-6`
+ DietPi-Update | Improved error handling and user output
  - Failsafe: Check all (core, sub and RC) version integers, as those are used in arithmetic and printed later.
+ DietPi-Update | Minor coding and wording
+ DietPi-Update | Reduce required free space to 100M
  - As the update itself does require nearly no disk space (RAMdisk + download/log to RAM), besides `/rootfs` files patching, which are very small, mostly just replaced. DietPi-Software and other script calls do their own space checks.
+ DietPi-Pre-Patch | Implement new pre-patch, that runs prior to incremental patch, APT updates and DietPi archive download
+ DietPi-Patch | Merge EMR patch system into DietPi-Pre-Patch
  - As well, the manual additional update execution renders pre v6.9 rootfs merge obsolete.
+ DietPi-Update | Add new DietPi-Pre-patch system to dietpi-update procedure
+ DietPi-Patch | Redo removal of obsolete DietPi files, if they were not removed from disk as well

MichaIng committed:
v6.17

+ DietPi-Update | Allow RC version updates on all branches
+ DietPi-Update | Improved error handling and user output
+ DietPi-Update | Minor coding and wording

MichaIng committed:
v6.17

+ DietPi-Update | Further coding and wording
+ DietPi-Update | Reduce required free space to 100M, as the update itself does require nearly no disk space (RAMdisk + download/log to RAM), besides /rootfs files patching, which are very small, mostly just replaced. DietPi-Software and other script calls do their own space checks.

MichaIng committed:
v6.17

+ DietPi-Pre-Patch | Implement new pre-patch, that runs prior to incremental patch, APT updates and DietPi archive download

MichaIng committed:
v6.17

+ DietPi-Patch | Move from EMR patch system to new DietPi-Pre-Patch system

MichaIng committed:
v6.17

+ DietPi-Update | Add new DietPi-Pre-patch system to dietpi-update procedure

MichaIng commented:
ToDo:
- `rm /DietPi/dietpi/...` alone does not remove obsolete files from DietPi system. `rm /boot/dietpi/...` needs to be done as well, otherwise `dietpi-ramdisk 0` will recover those files from disk.

Question:
- Move update system switch into `pre-patch_file`, just running `[[ -f /DietPi/dietpi/pre-patch_file ]] && ./DietPi/dietpi/pre-patch_file; rm /DietPi/dietpi/pre-patch_file` inside `patch_file`?

Fourdee commented:
@MichaIng 

>  rm /boot/dietpi/... 

Must be avoided. Due to: https://github.com/Fourdee/DietPi/issues/905#issuecomment-298241622

Unless we add sanity checks and error checking with stable recovery, before and after removal. 

Ideally, all removals of obsolete scripts should be done during ```patch_file```

MichaIng commented:
@Fourdee
> Must be avoided. Due to: #905 (comment)

I not mean to remove all content of /boot/dietpi, just the obsolete scripts we do actively remove from ramdisk during patch due to moved to other place, merged into other script or removed for other reasons.

E.g. my v6.16 system still contains dietpi-ramdisk/ramlog/logclear in /DietPi/dietpi, as recovered on shutdown after update.
So removal during patch  need to be done in both, ramdisk and disk.
____
`G_DIETPI-NOTIFY 4 "$G_PROGRAM_NAME" 'Completed'`
Was meant to be NOTIFY 3, right? :smile:

MichaIng committed:
Merge branch 'dev' into dietpi-update

MichaIng commented:
Quick test:
```
rm /DietiPi/dietpi/dietpi-ramdisk
reboot
...
root@VM-Stretch:~# l /DietPi/dietpi/dietpi-ramdisk
-rwxr-xr-x 1 root root 4451 Oct  9 16:14 /DietPi/dietpi/dietpi-ramdisk
```
- I will move all the removals to v6.17 patch and remove from disk as well: https://github.com/Fourdee/DietPi/pull/2123/commits/1e7c0eb35d6d9684734c3c4b41e0a46af2d1d0de

MichaIng committed:
v6.17

+ DietPi-Patch | Redo removal of obsolete DietPi files, if they were not removed from disk as well

MichaIng committed:
v6.17

+ DietPi-Update | Fix missing pre-patch_file exec permissions

MichaIng committed:
v6.17

DietPi-Update | Little workaround, in case .version was not created successfully and for easier testing of update system

Fourdee commented on dietpi/dietpi-update:
```diff
@@ -238,11 +241,11 @@ Please select 'Update' option to apply the update."
 ------------------------Notice------------------------
-The benefit of DietPi is we use standard linux configurations and commands. The downside is we can't possibly accommodate or predict, every modifcation to Linux configurations files by the end user, during the update.
-Although we test the updates thoroughly, if you have made custom changes to Linux configuration files, outside of the DietPi programs, an update may trigger a potential issue.
+A benefit of DietPi is, that we use standard Linux (Debian) configurations and commands. The downside is, that we can't possibly accommodate or predict every modifcation to Linux configurations files by the end user, during the update.
```
Doesn't work for me 😄 

You could do
```
- A benefit of DietPi is: We use standard Linux (Debian) configurations and commands.
- A potential downside is: We can't possibly accommodate or predict all modification to Linux configurations files by the end user, outside of DietPi programs, during updates.
```

Fourdee replied:
I'll make the changes.

MichaIng replied:
@Fourdee 
Checking `deepl.com`, yes you are right, the additional `, that` made it worse 😅.
Actually I just found `The benefit` a bid too pronounced, since we have so much benefits.
Your solution looks more structured, nice as well 👍.

Fourdee committed:
Minor wording

Fourdee commented:
@MichaIng 

Looks good, many thanks 👍 

Merged for additional testing.

Fourdee commented:
@MichaIng 

Quick test, ```pre-patch_file``` does not run, until ```dietpi-update``` is finished and updated again. Need to add a manual run for this, pre ```dietpi-update```.

MichaIng commented:
@Fourdee 
Should be actually okay, since it is the very first step of second update run.

I first tried to implement it the other way round:
- Implement new versioning system into pre-patch_file.
- Call pre-patch as only left "header" patch in patch_file.

But the additional download, where first Git branch and owner would be needed again (🤔 actually no download needed, hmm don't remember any more exactly^^), the double nested script call and such, made me solve it like this. There is no chance to apply special patches earlier then in incremental patch stage anyway (for pre v6.17 systems). Doing it then or as a very first step of next update run for my impression does not make much difference. Should be okay as it is.
____________
Hmm looking again on it, actually it would look somehow cleaner to do pre-v6.17 pre-patch call in patch_file and RC versioning implementation inside pre-patch then. Not sure what made me switch to current solution....

Fourdee commented:
@MichaIng 

> actually it would look somehow cleaner to do pre-v6.17 pre-patch call in patch_file and RC versioning implementation inside pre-patch then

To be honest, until we update images with the new ```dietpi-update``` and ```pre-patch_file``` system, we would need to manually run the pre-patches in ```patch_file```, duped regardless.

Suppose we could run ```pre-patch_file``` via ```patch_file``` during v6.17 update once. After which new system will be used.

Fourdee commented:
https://github.com/Fourdee/DietPi/commit/e6a5761af947f384d02e5540d5381cb042e40c91 should do it, needs testing.

Fourdee commented:
@MichaIng 

Yep you were right, not required due to forced re-run of ```dietpi-update``` to enable new systems.
